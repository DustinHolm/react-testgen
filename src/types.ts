export interface ImportBlock {
    sourceFile: string,
    isInternal: boolean,
    imports: Import[]
}

export interface Import {
    isDefault: boolean,
    name: string,
    element?: ExportableElement
}

export interface Export {
    name: string,
    isDefault: boolean,
    element?: ExportableElement
}

export enum ExportType {
    Class,
    Function,
    Interface,
    Variable
}

export interface ExportableElement {
    type: ExportType,
    returnsJSX: boolean
}

export interface ClassElement extends ExportableElement {
    methods: MethodElement[],
    constructors: Constructor[]
}

export interface MethodElement extends FunctionElement {
    name: string
}

export interface FunctionElement extends ExportableElement {
    returnType: string,
    parameters: Attribute[]
}

export interface InterfaceElement extends ExportableElement {
    attributes: Attribute[]
}

export interface VariableElement extends ExportableElement {
    varType: string,
    varValue: string
}

export interface Attribute {
    name: string,
    type: string
}

export interface Constructor {
    parameters: Attribute[]
}