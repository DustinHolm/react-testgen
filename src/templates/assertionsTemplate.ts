import { Export, ExportType, FunctionElement } from "../types";
import { getMockName, toPascalCase } from "./common";

export const createAssertion = (mock: Export): string => {
    if (mock.element?.returnsJSX === true) {
        return createJSXAssertion(mock)
    } else if (mock.element?.type === ExportType.Function) {
        return createFunctionAssertion(mock)
    } else {
      return ""
    }
}

const createJSXAssertion = (mock: Export): string => {
    const name = toPascalCase(mock.name)
    const mockName = getMockName(mock.name)
    const testId = mockName

    return `\
function then${name}IsRendered(expected: boolean) {
    if (expected) {
        expect(screen.getByTestId("${testId}")).toBeInTheDocument()
    } else {
        expect(screen.queryByTestId("${testId}")).toBeNull()
    }
}

function then${name}WasCalledWith(props: any) {
    expect(${mockName}).toHaveBeenCalledWith([props])
}
`
}

const createFunctionAssertion = (mock: Export): string => {
    const name = toPascalCase(mock.name)
    const mockName = getMockName(mock.name)
    const parameters = (mock.element as FunctionElement).parameters

    if (parameters.length === 0) {
        return `
function then${name}WasCalled(expected: boolean) {
    if (expected) {
        expect(${mockName}).toHaveBeenCalled()
    } else {
        expect(${mockName}).not.toHaveBeenCalled()
    }
}
`
    }

    const parameterString = parameters.map(p => 
      `${p.name}: ${p.type}`
      ).join(", ")
    const parameterCallString = parameters.map(p => p.name).join(", ")

    return `\
function then${name}WasCalledWith(${parameterString}) {
    expect(${mockName}).toHaveBeenCalledWith([${parameterCallString}])
}
`
}
