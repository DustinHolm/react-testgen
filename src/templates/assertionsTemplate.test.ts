import { Export, ExportType, FunctionElement, ImportBlock } from "../types";
import { createAssertion } from "./assertionsTemplate";

describe("assertionsTemplate Tests", () => {
    test("For React component", () => {
        const element: FunctionElement = {
            returnsJSX: true,
            type: ExportType.Function,
            returnType: "JSX.Element",
            parameters: []
        }
        const exportElement: Export = {name: "MockComponent", isDefault: true, element: element}
        const returnedString = createAssertion(exportElement)
        expect(returnedString).toMatchSnapshot()
    })

    test("For function without parameters", () => {
        const element: FunctionElement = {
            returnsJSX: false,
            type: ExportType.Function,
            returnType: "number",
            parameters: []
        }
        const exportElement: Export = {name: "MockComponent", isDefault: true, element: element}
        const returnedString = createAssertion(exportElement)
        expect(returnedString).toMatchSnapshot()
    })

    test("For function with parameters", () => {
        const element: FunctionElement = {
            returnsJSX: false,
            type: ExportType.Function,
            returnType: "number",
            parameters: [{name: "num", type: "number"}, {name: "str", type: "string"}]
        }
        const exportElement: Export = {name: "MockComponent", isDefault: true, element: element}
        const returnedString = createAssertion(exportElement)
        expect(returnedString).toMatchSnapshot()
    })
})