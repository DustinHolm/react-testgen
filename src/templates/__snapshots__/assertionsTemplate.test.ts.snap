// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertionsTemplate Tests For React component 1`] = `
"function thenMockComponentIsRendered(expected: boolean) {
    if (expected) {
        expect(screen.getByTestId(\\"mockMockComponent\\")).toBeInTheDocument()
    } else {
        expect(screen.queryByTestId(\\"mockMockComponent\\")).toBeNull()
    }
}

function thenMockComponentWasCalledWith(props: any) {
    expect(mockMockComponent).toHaveBeenCalledWith([props])
}
"
`;

exports[`assertionsTemplate Tests For function with parameters 1`] = `
"function thenMockComponentWasCalledWith(num: number, str: string) {
    expect(mockMockComponent).toHaveBeenCalledWith([num, str])
}
"
`;

exports[`assertionsTemplate Tests For function without parameters 1`] = `
"
function thenMockComponentWasCalled(expected: boolean) {
    if (expected) {
        expect(mockMockComponent).toHaveBeenCalled()
    } else {
        expect(mockMockComponent).not.toHaveBeenCalled()
    }
}
"
`;
